### PR TITLE
Forward aspire run arguments to AppHost

### PR DESCRIPTION
**Pull request by sebastienros**
## Description

`aspire run -- <args>` should pass those arguments through to the AppHost process so AppHosts can read them from their normal process argument APIs. This change preserves `System.CommandLine` unmatched tokens from the command separator and threads them through C# AppHost launches, guest-language AppHost launches, detached `run`/`start` child invocations, and extension debug-session delegation.

The forwarding path adds the `--` separator only when reconstructing a child CLI invocation, while preserving literal `--` values that are part of the AppHost arguments. Coverage includes focused CLI unit tests for parser behavior, C# `dotnet run` argument placement, extension delegation, guest runtime execution, and a TypeScript CLI E2E test that verifies a guest `apphost.ts` receives forwarded arguments.

## Checklist

- Is this feature complete?
  - [x] Yes. Ready to ship.
  - [ ] No. Follow-up changes expected.
- Are you including unit tests for the changes and scenario tests if relevant?
  - [x] Yes
  - [ ] No
- Did you add public API?
  - [ ] Yes
    - If yes, did you have an API Review for it?
      - [ ] Yes
      - [ ] No
    - Did you add `<remarks />` and `<code />` elements on your triple slash comments?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change make any security assumptions or guarantees?
  - [ ] Yes
    - If yes, have you done a threat model and had a security review?
      - [ ] Yes
      - [ ] No
  - [x] No
- Does the change require an update in our Aspire docs?
  - [ ] Yes
    - Link to `aspire.dev` issue:
      - [New issue](https://github.com/microsoft/aspire.dev/issues/new)
  - [x] No

### CODE REVIEW

File: src/Aspire.Cli/Projects/GuestRuntime.cs [hunk 1]
```diff
@@ -135,12 +135,35 @@ public GuestRuntime(RuntimeSpec spec, ILogger logger, FileLoggerProvider? fileLo
...
+        string[]? runArgs,
+        CancellationToken cancellationToken)
     {
         var commandSpec = watchMode && _spec.WatchExecute is not null
             ? _spec.WatchExecute
             : _spec.Execute;
 
-        return await ExecuteCommandAsync(commandSpec, appHostFile, directory, environmentVariables, null, launcher, cancellationToken);
+        return await ExecuteCommandAsync(commandSpec, appHostFile, directory, environmentVariables, runArgs, launcher, cancellationToken);
     }
```
[the remark this answers is not in the document]
Comment (sebastienros) [reply src/Aspire.Cli/Projects/GuestRuntime.cs]: Good catch. I updated exact `{args}` placeholder handling to expand forwarded arguments into separate argv entries, and added coverage for both run and publish command specs using that placeholder.

**Review by IEvangelist — APPROVED** [review]

**Comment by davidfowl** [comment]
Needs pr-testing report

**Comment by IEvangelist** [comment]
Code-review only — CI on the current head doesn't have fresh artifacts, so I read the diff.

Walked the forwarding mechanism in `AppHostLauncher.GetAppHostArguments` / `AddAppHostArguments` plus the new `{args}` placeholder path in `GuestRuntime.ReplacePlaceholders` and chased it through the language-specific RuntimeSpecs:

- **TypeScript**: spec is `["--no-install", "tsx", "--tsconfig", AppHostTsConfigFileName, "{appHostFile}"]` (no `{args}` token) — the fallback at the end of `ReplacePlaceholders` (`if (additionalArgs is { Length: > 0 } && !args.Any(a => a.Contains("{args}")))`) appends args as separate ArgumentList elements. ✅ correct argument forwarding for TS.
- **Python**: standalone `"{args}"` element — goes through `result.AddRange(additionalArgs)`. ✅ array semantics preserved.
- **Java**: this is the one I want to flag.

### One concern — embedded `{args}` in the Java RuntimeSpec is shell-injectable

In `src/Aspire.Hosting.CodeGeneration.Java/JavaLanguageSupport.cs` (pre-existing — not changed by this PR, but this PR is what makes user arguments reach it via `aspire run -- <args>`):

```csharp
? ["/c", "if not exist .java-build mkdir .java-build && javac --enable-preview --source 25 -d .java-build @.modules\\sources.txt AppHost.java && java --enable-preview -cp .java-build AppHost {args}"]
: ["-c", "mkdir -p .java-build && javac --enable-preview --source 25 -d .java-build @.modules/sources.txt AppHost.java && java --enable-preview -cp .java-build AppHost {args}"]
```

The command is `cmd` (Windows) / `sh` (Unix), and the `{args}` token is embedded inside the `/c` / `-c` payload — i.e. the script the shell is asked to interpret. When `ReplacePlaceholders` hits the embedded path it joins args with a single space:

```csharp
if (replaced.Contains("{args}"))
{
    if (additionalArgs is { Length: > 0 })
    {
        replaced = replaced.Replace("{args}", string.Join(" ", additionalArgs));
    }
    ...
}
```

So with this PR's `aspire run -- "foo; rm -rf ~/important"` on a Java AppHost, the spawned shell will execute the `rm -rf ~/important` part because the user's argument lands inside the shell command line. `UseShellExecute=false` only stops *the parent CLI* from going through a shell — it doesn't help when the command we explicitly invoke *is itself a shell*. This is more "user shoots themselves in the foot" than RCE in most cases (the user owns the args), but in CI where `aspire run -- $GITHUB_INPUTS` or similar is conceivable, untrusted input would execute as shell commands. And it's also a correctness regression for any Java apphost user who legitimately wants to pass multi-token argv with spaces.

### Suggested fix options

A few directions, in rough order of preference:

1. **Reject embedded `{args}` in `ReplacePlaceholders` outright.** Easiest, safest, and matches what the Python / TS / Go specs already do (standalone token only). Add a clear error message and update the Java spec to follow the standalone form. The Java spec would need to drop the `cmd /c` / `sh -c` wrapping and use a separate launcher step for the conditional `mkdir` so the final `java` invocation can be a plain `ArgumentList` entry with `{args}` as a standalone token.
2. **Quote-escape the joined args** in the embedded path (`ShellArgumentQuote` per platform). Less appealing — quoting rules differ between `cmd` and `POSIX sh` and getting this right is famously error-prone.
3. **Drop the embedded path entirely.** Same as (1) but more aggressive — at minimum, the documented "Supports placeholders: ... {args}" comment on `RuntimeSpec.cs` should specify "standalone token only".

A test like `GuestRuntimeTests.ReplacePlaceholders_EmbeddedArgsWithSpacedToken_DoesNotLoseBoundaries` would catch this for any future spec author.

### Minor: arg logging at Debug

`_logger.LogDebug("Launching: {Command} {Args}", commandSpec.Command, string.Join(" ", args));` — fine at Debug, but worth a one-liner in the docs (or a comment near the log) noting that user-provided args end up in trace logs so secrets shouldn't be passed via `--`. Not blocking.

### Everything else

The `AppHostLauncher.GetAppHostArguments` / `AddAppHostArguments` helpers, the `UnmatchedTokens` plumbing through `DebugSessionOptions`, the literal-`--` preservation test, and the new E2E `TypeScriptEmptyAppHostTemplate` argument test all look correct. `System.CommandLine`'s `--` boundary is respected and args after it land in `UnmatchedTokens` cleanly.

LGTM modulo the Java embedded-`{args}` follow-through (which I'd really like to see fixed before merge given this PR is what makes user input reach that shell line).

**Comment by sebastienros** [comment]
> Needs pr-testing report

PR testing report:
- `dotnet test --project tests/Aspire.Cli.Tests/Aspire.Cli.Tests.csproj --no-launch-profile -- --filter-class "*.RunCommandTests" --filter-class "*.StartCommandTests" --filter-class "*.DotNetCliRunnerTests" --filter-class "*.GuestRuntimeTests" --filter-not-trait "quarantined=true" --filter-not-trait "outerloop=true"` — passed, 151 tests.
- `dotnet test --project tests/Aspire.Hosting.CodeGeneration.Java.Tests/Aspire.Hosting.CodeGeneration.Java.Tests.csproj --no-launch-profile -- --filter-not-trait "quarantined=true" --filter-not-trait "outerloop=true"` — passed, 24 tests.
- `dotnet test --project tests/Aspire.Cli.EndToEnd.Tests/Aspire.Cli.EndToEnd.Tests.csproj --no-launch-profile -- --filter-method "*.RunTypeScriptEmptyAppHost_ForwardsArgumentsToAppHostProcess" --filter-not-trait "quarantined=true" --filter-not-trait "outerloop=true"` — attempted locally, but the Docker-based E2E could not start because the local Docker daemon is unavailable. The previous PR CI recording for `RunTypeScriptEmptyAppHost_ForwardsArgumentsToAppHostProcess` passed, and the current push will let CI re-run it on PR infrastructure.

> One concern — embedded `{args}` in the Java RuntimeSpec is shell-injectable

Fixed in `5f512f35c`. Java AppHosts now compile in `PreExecute` with direct `javac`, then run with direct `java` and standalone `{args}` so forwarded arguments stay separate process arguments and never get interpolated into a shell payload. `GuestRuntime` also now rejects embedded `{args}` placeholders up front, with tests covering argument-boundary preservation and embedded-placeholder rejection.

**Comment by sebastienros** [comment]
# PR Testing Report

## PR Information
- **PR Number:** #16630
- **Title:** Forward aspire run arguments to AppHost
- **Head Commit:** 8c264cc40f6a6be1850862000f42c68ca8dc1113
- **Tested At:** 2026-06-23T16:07:00Z

## Artifact Version Verification
- **Expected Commit:** 8c264cc4 (PR head)
- **Installed Version:** `13.5.0-pr.16630.g8c264cc4`
- **Status:** ✅ Verified (version contains PR head short SHA `g8c264cc4`)

## Changes Analyzed
### Files Changed
- `src/Aspire.Cli/Commands/AppHostLauncher.cs` — arg-forwarding helpers
- `src/Aspire.Cli/Commands/RunCommand.cs` — `aspire run` forwards `-- args`
- `src/Aspire.Cli/Commands/StartCommand.cs` — `aspire start` forwards `-- args`
- `src/Aspire.Cli/Projects/GuestAppHostProject.cs` — guest arg threading
- `src/Aspire.Cli/Projects/GuestRuntime.cs` — `{args}` placeholder expansion
- `src/Aspire.Hosting.CodeGeneration.Java/JavaLanguageSupport.cs` — Java codegen no-shell-expansion fix
- `src/Aspire.TypeSystem/RuntimeSpec.cs` — `{args}` placeholder support
- Tests: `TypeScriptEmptyAppHostTemplateTests.cs`, `RunCommandTests.cs`, `StartCommandTests.cs`, `GuestRuntimeTests.cs`, `JavaLanguageSupportTests.cs`

### Change Categories
- [x] CLI changes detected — `aspire run`/`aspire start` forward args after `--` to AppHost
- [x] Hosting changes — Java codegen security fix; `{args}` placeholder in TypeSystem
- [ ] Dashboard changes
- [ ] Template changes
- [ ] VS Code extension changes
- [x] Test changes

## Test Environment
- **Execution target:** Local (macOS arm64), temp dir `/var/folders/qn/lq74424935j69rxktvp5tqwc0000gn/T/aspire-pr-test-XXXXXX.Lgyrk2w6QT`
- **Node.js:** v24.4.1 (for the TypeScript guest scenario)
- **CLI:** dogfood PR build, isolated install

## Test Scenarios Executed

### Scenario 1: C# AppHost arg forwarding (happy path)
**Objective:** Verify `aspire run -- arg1=value1 --flag` reaches the C# AppHost `string[] args`.
**Coverage Type:** Happy path
**Status:** ✅ Passed

**Steps:**
1. `aspire new aspire-empty --language csharp` (file-based `apphost.cs`).
2. Edited `apphost.cs` to write received `args` to a file before `CreateBuilder(args)`.
3. Ran `aspire run -- arg1=value1 --flag`.

**Result:**
```
COUNT=2
ARG[0]=<arg1=value1>
ARG[1]=<--flag>
```
Both tokens forwarded exactly, in order.

---

### Scenario 2: C# arg boundary & shell-safety (boundary / unhappy path)
**Objective:** Verify args with spaces and shell metacharacters arrive intact as separate tokens, with no re-splitting or shell evaluation.
**Coverage Type:** Boundary / security
**Status:** ✅ Passed

**Steps:**
1. Ran `aspire run -- "hello world" "; echo pwned" "--name=a b"` against the same AppHost.

**Result:**
```
COUNT=3
ARG[0]=<hello world>
ARG[1]=<; echo pwned>
ARG[2]=<--name=a b>
```
Spaces preserved; `; echo pwned` delivered literally (no shell injection); each arg kept as one token.

---

### Scenario 3: TypeScript guest AppHost forwarding (happy path)
**Objective:** Verify forwarded args reach a guest (Node.js/TypeScript) AppHost via the `{args}` placeholder.
**Coverage Type:** Happy path (guest/polyglot)
**Status:** ✅ Passed

**Steps:**
1. `aspire new aspire-empty --language typescript` (`apphost.mts`, deps installed).
2. Edited `apphost.mts` to write `process.argv.slice(2)` to a file.
3. Ran `aspire run -- foo=bar "; echo pwned" "a b c"`.

**Result:**
```
COUNT=3
ARG[0]=<foo=bar>
ARG[1]=<; echo pwned>
ARG[2]=<a b c>
```
Args forwarded to the running node process; boundaries preserved; no shell injection. Dashboard started normally, confirming a genuinely running guest AppHost.

---

### Scenario 4: No-`--` regression (negative)
**Objective:** Verify plain `aspire run` (no `--`) starts cleanly and forwards no extra args.
**Coverage Type:** Negative / regression
**Status:** ✅ Passed

**Steps:**
1. Ran `aspire run` (no `--`) against the C# AppHost.

**Result:**
```
COUNT=0
```
No spurious args injected; AppHost started normally.

## Summary
| Scenario | Status | Notes |
|----------|--------|-------|
| 1. C# arg forwarding | ✅ Passed | `arg1=value1`, `--flag` forwarded exactly |
| 2. C# boundary / shell-safety | ✅ Passed | Spaces + `; echo pwned` intact, no injection |
| 3. TypeScript guest forwarding | ✅ Passed | `{args}` placeholder forwards to node, no injection |
| 4. No-`--` regression | ✅ Passed | COUNT=0, no breakage |

## Overall Result
**✅ PR VERIFIED**

### Recommendations
- Arg forwarding works end-to-end for both C# (file-based AppHost) and a TypeScript guest AppHost, including correct argument-boundary preservation and no shell injection for shell-metacharacter inputs — matching the intent of the Java codegen security fix.
- Verified on macOS arm64 only. TypeScript E2E for this behavior also runs in CI (`TypeScriptEmptyAppHostTemplateTests`); other guest runtimes (e.g. Java, Python) require their toolchains and are covered by the polyglot SDK tests rather than this local run.

File: src/Aspire.Hosting.CodeGeneration.Java/JavaLanguageSupport.cs [hunk 2]
```diff
@@ -102,16 +102,22 @@ public RuntimeSpec GetRuntimeSpec()
...
             InstallDependencies = null,
+            PreExecute =
+            [
+                new CommandSpec
+                {
+                    Command = "javac",
+                    Args = OperatingSystem.IsWindows()
+                        ? ["--enable-preview", "--source", "25", "-d", ".java-build", "@.modules\\sources.txt", "AppHost.java"]
+                        : ["--enable-preview", "--source", "25", "-d", ".java-build", "@.modules/sources.txt", "AppHost.java"]
+                }
```
[the remark this answers is not in the document]
Comment (sebastienros) [reply src/Aspire.Hosting.CodeGeneration.Java/JavaLanguageSupport.cs:115]: Good catch, accepted. The generator (`AtsJavaCodeGenerator`) writes `sources.txt` under `.aspire/modules/` and its contents reference `.aspire/modules/...`, but `javac` was pointed at `@.modules/sources.txt`, so compilation would not find the argument file. I fixed the path to `@.aspire/modules/sources.txt` (and the Windows `@.aspire\modules\sources.txt` variant) and tightened the test to assert the exact PreExecute `javac` args.

File: tests/Aspire.Cli.EndToEnd.Tests/TypeScriptEmptyAppHostTemplateTests.cs [hunk 3]
```diff
@@ -170,4 +171,70 @@ private static string GetSdkVersion(string appDirectory)
...
+        var strategy = CliInstallStrategy.Detect(output.WriteLine);
+        var workspace = TemporaryWorkspace.Create(output);
+
+        using var terminal = CliE2ETestHelpers.CreateDockerTestTerminal(repoRoot, strategy, output, mountDockerSocket: true, workspace: workspace);
+
+        var pendingRun = terminal.RunAsync(TestContext.Current.CancellationToken);
+
+        var counter = new SequenceCounter();
+        var auto = new Hex1bTerminalAutomator(terminal, defaultTimeout: TimeSpan.FromSeconds(500));
+
```
[the remark this answers is not in the document]
Comment (sebastienros) [reply tests/Aspire.Cli.EndToEnd.Tests/TypeScriptEmptyAppHostTemplateTests.cs]: Accepted. I updated the test to wrap the run in `TerminalRun` via `CliE2ETestHelpers.StartRun(...)` with `await using`, matching the established pattern in the sibling template tests, so the terminal is torn down and diagnostics are captured even if an assertion throws before the explicit exit.

**Comment by davidfowl** [comment]
@sebastienros closing this

**Comment by sebastienros** [comment]
@davidfowl you don't like it?
